File: backend/app/services/leave_hours.py
"""請假時數的純函式。

與出勤工時（work_hours.py）刻意不同的兩點，都是規則本身的差異、不是不一致：

1. **請假的午休固定不浮動**：請假通常事前申請，當天沒有打卡時間可據以浮動，
   硬要浮動只會讓基準比規則本身更難解釋。
2. **緩衝只套用在整段請假區間真正的頭尾兩端**：多天請假的中間日、以及非頭尾
   那一側，一律用未緩衝的表定時間。否則「週五 09:00 到週一 18:00」這種剛好卡
   整點的多天請假，會因為兩端各自誤套緩衝而多算 20 分鐘（16.00 變成 16.33）。
"""

from datetime import date, datetime, timedelta
from zoneinfo import ZoneInfo

from app.services.work_hours import WorkSettings, at


def _local_date(moment: datetime, tz: str) -> date:
    return moment.astimezone(ZoneInfo(tz)).date()


def calculate_leave_hours_by_day(
    start_time: datetime,
    end_time: datetime,
    settings: WorkSettings,
    tz: str,
    holiday_dates: set[date] | None = None,
) -> dict[date, float]:
    """把請假區間依台北時區逐日切分，跳過週末與國定假日，回傳 {日期: 當日時數}。

    每個工作日的「工作時間窗」切成 `[work_start − grace, lunch_start]` 與
    `[lunch_end, work_end + grace]` 兩段分別取交集後相加——不是拿整段
    work_start~work_end 直接相減，那會把午休也算成請假時數。

    刻意不在這裡四捨五入：呼叫端要先加總原始值再一次四捨五入，逐日先各自
    進位再加總會累積誤差。
    """
    holiday_dates = holiday_dates or set()

    first_day = _local_date(start_time, tz)
    last_day = _local_date(end_time, tz)

    by_day: dict[date, float] = {}
    cursor = first_day

    while cursor <= last_day:
        if cursor.weekday() < 5 and cursor not in holiday_dates:
            start_grace = settings.grace if cursor == first_day else timedelta(0)
            end_grace = settings.grace if cursor == last_day else timedelta(0)

            day_work_start = at(cursor, settings.work_start, tz) - start_grace
            day_work_end = at(cursor, settings.work_end, tz) + end_grace
            day_lunch_start = at(cursor, settings.lunch_start, tz)
            day_lunch_end = at(cursor, settings.lunch_end, tz)

            day_total = timedelta(0)
            for segment_start, segment_end in (
                (day_work_start, day_lunch_start),
                (day_lunch_end, day_work_end),
            ):
                lo = max(start_time, segment_start)
                hi = min(end_time, segment_end)
                if hi > lo:
                    day_total += hi - lo

            if day_total > timedelta(0):
                by_day[cursor] = day_total.total_seconds() / 3600

        cursor += timedelta(days=1)

    return by_day


def compute_leave_hours_for_date(
    punch_date: date,
    leave_intervals,
    settings: WorkSettings,
    tz: str,
) -> float:
    """加總當日所有已核准請假區間與表定工時的交集時數。

    沿用 calculate_leave_hours_by_day() 的逐日交集邏輯，確保出勤這一側算出的
    「當日已請假時數」與請假申請單自己算出的時數規則永遠同步。
    呼叫端已保證 punch_date 是工作日，故這裡不需要再傳 holiday_dates。
    """
    total = 0.0
    for start, end in leave_intervals:
        by_day = calculate_leave_hours_by_day(start, end, settings, tz, holiday_dates=set())
        total += by_day.get(punch_date, 0.0)
    return total

File: frontend/src/utils/leaveHours.js
// 請假時數的前端本地預覽（UI-SPEC.md §3.8「即時預估時數」）。
//
// 演算法必須跟後端 backend/app/services/leave_hours.py 的 calculate_leave_hours_by_day()
// 逐步對應：逐工作日切分、排除週末與國定假日、緩衝只套用在整段請假區間真正的
// 頭尾兩端、午休固定不浮動。這裡只是預覽，正式時數以送出後的後端回應為準，但
// 演算法本身不能各算一套，否則預覽數字會誤導使用者。

const TAIPEI_OFFSET = "+08:00";

function combine(dateStr, timeStr) {
  // timeStr 可能是 "HH:mm" 或後端設定值的 "HH:mm:ss"，統一只取到分鐘。
  const [hour, minute] = timeStr.split(":");
  return new Date(`${dateStr}T${hour}:${minute}:00${TAIPEI_OFFSET}`);
}

function addMinutes(date, minutes) {
  return new Date(date.getTime() + minutes * 60_000);
}

// 用 Date.UTC 而非瀏覽器本地時區建構，星期幾是曆法日期本身的性質、跟時區無關，
// 避免瀏覽器所在時區在日期邊界附近算錯。
function dayOfWeek(dateStr) {
  const [year, month, day] = dateStr.split("-").map(Number);
  return new Date(Date.UTC(year, month - 1, day)).getUTCDay(); // 0=Sun ... 6=Sat
}

function nextDateString(dateStr) {
  const [year, month, day] = dateStr.split("-").map(Number);
  return new Date(Date.UTC(year, month - 1, day + 1)).toISOString().slice(0, 10);
}

/**
 * @param {object} params
 * @param {string} params.startDate 'YYYY-MM-DD'
 * @param {string} params.startTime 'HH:mm'
 * @param {string} params.endDate 'YYYY-MM-DD'
 * @param {string} params.endTime 'HH:mm'
 * @param {{work_start_time:string, work_end_time:string, lunch_start_time:string, lunch_end_time:string, grace_period_minutes:number}} params.settings
 * @param {Set<string>} [params.holidayDates] 'YYYY-MM-DD' 集合
 * @returns {number} 預估時數，四捨五入至小數點後兩位；輸入不完整或區間無效時回傳 0
 */
export function estimateLeaveHours({ startDate, startTime, endDate, endTime, settings, holidayDates }) {
  if (!startDate || !startTime || !endDate || !endTime || !settings || endDate < startDate) return 0;

  const start = combine(startDate, startTime);
  const end = combine(endDate, endTime);
  if (!(end > start)) return 0;

  const grace = settings.grace_period_minutes ?? 0;
  const holidays = holidayDates ?? new Set();

  let totalMs = 0;
  let cursor = startDate;
  while (cursor <= endDate) {
    const dow = dayOfWeek(cursor);
    if (dow !== 0 && dow !== 6 && !holidays.has(cursor)) {
      const startGrace = cursor === startDate ? grace : 0;
      const endGrace = cursor === endDate ? grace : 0;

      const dayWorkStart = addMinutes(combine(cursor, settings.work_start_time), -startGrace);
      const dayWorkEnd = addMinutes(combine(cursor, settings.work_end_time), endGrace);
      const dayLunchStart = combine(cursor, settings.lunch_start_time);
      const dayLunchEnd = combine(cursor, settings.lunch_end_time);

      for (const [segmentStart, segmentEnd] of [
        [dayWorkStart, dayLunchStart],
        [dayLunchEnd, dayWorkEnd],
      ]) {
        const lo = start > segmentStart ? start : segmentStart;
        const hi = end < segmentEnd ? end : segmentEnd;
        if (hi > lo) totalMs += hi.getTime() - lo.getTime();
      }
    }
    cursor = nextDateString(cursor);
  }

  return Math.round((totalMs / 3_600_000) * 100) / 100;
}
